engine/table/src/main/java/io/deephaven/engine/table/impl/hierarchical/RollupNodeKeySource.java

```diff
@@ -0,0 +1,183 @@
+    private final ColumnSource<?>[] groupByValueSources;
+
+    /**
+     * Construct a new RollupNodeKeySource backed by the supplied column sources.
+     *
+     * @param depthSource A source of {@code int} depths that determines how many of the {@code groupByValueSources}
+     *        should have their values included in the result
+     * @param groupByValueSources Sources corresponding to the aggregation group-by values in the rollup
+     */
+    RollupNodeKeySource(
+            @NotNull final ColumnSource<Integer> depthSource,
+            @NotNull final ColumnSource<?>... groupByValueSources) {
+        this.depthSource = depthSource;
+        this.groupByValueSources = Stream.of(groupByValueSources)
+                .map(ReinterpretUtils::maybeConvertToPrimitive).toArray(ColumnSource[]::new);
+    }
+
+    @Override
+    public ChunkType getChunkType() {
+        return ChunkType.Object;
+    }
+
+    @Override
+    public void fillChunk(
+            @NotNull final ChunkSource.FillContext context,
+            @NotNull final WritableChunk<? super Values> destination,
+            @NotNull final RowSequence rowSequence) {
+        fillChunkInternal(context, destination, rowSequence, false);
+    }
+
+    public void fillPrevChunk(
+            @NotNull final ChunkSource.FillContext context,
+            @NotNull final WritableChunk<? super Values> destination,
+            @NotNull final RowSequence rowSequence) {
+        fillChunkInternal(context, destination, rowSequence, true);
+    }
+
+    private void fillChunkInternal(
+            @NotNull final ChunkSource.FillContext context,
+            @NotNull final WritableChunk<? super Values> destination,
+            @NotNull final RowSequence rowSequence,
+            final boolean usePrev) {
+        if (rowSequence.isEmpty()) {
+            destination.setSize(0);
+            return;
+        }
+        final FillContext fc = (FillContext) context;
+        final IntChunk<? extends Values> depths = usePrev
+                ? depthSource.getPrevChunk(fc.depthContext, rowSequence).asIntChunk()
+                : depthSource.getChunk(fc.depthContext, rowSequence).asIntChunk();
+        final int maxKeyWidth = getMaxKeyWidth(depths);
+        if (maxKeyWidth > groupByValueSources.length) {
+            throw new IllegalArgumentException(String.format("Invalid key width %d, maximum for this rollup is %d",
+                    maxKeyWidth, groupByValueSources.length));
+        }
+        final ObjectChunk<?, ? extends Values>[] groupByValues =
+                getGroupByValuesChunks(fc, rowSequence, usePrev, maxKeyWidth);
+        fillFromGroupByValues(rowSequence, depths, groupByValues, destination.asWritableObjectChunk());
+    }
+
+    private static int getMaxKeyWidth(@NotNull final IntChunk<? extends Values> depths) {
+        final int size = depths.size();
+        int maxKeyWidth = 0;
+        for (int kwi = 0; kwi < size; ++kwi) {
```

Comment:
I won't be able to get the actual key value, since I won't be able to fetch from the unknown column sources implied by the wider key. This problem also has to be handled here in the node key source; if we return a truncated node key, the information is lost, and we're probably doing the wrong thing.

I've added a short-circuit for the max key width iteration, and now I validate parent depth in the second, necessarily complete iteration.